pkg/license/license.go

```diff
@@ -5,18 +5,71 @@ import (
 	"fmt"
 	"io/ioutil"
 	"net/http"
+	"strings"
 
+	"github.com/gosimple/slug"
 	"github.com/pkg/errors"
 	kotsv1beta1 "github.com/replicatedhq/kots/kotskinds/apis/kots/v1beta1"
 	"github.com/replicatedhq/kots/kotskinds/client/kotsclientset/scheme"
 	"github.com/replicatedhq/kots/pkg/buildversion"
+	"github.com/replicatedhq/kots/pkg/logger"
+	"github.com/replicatedhq/kots/pkg/store"
 )
 
 type LicenseData struct {
 	LicenseBytes []byte
 	License      *kotsv1beta1.License
 }
 
+func ResolveExistingLicense(verifiedLicense *kotsv1beta1.License) (bool, error) {
+	installedAppSet := make(map[string]bool)
+
+	apps, err := store.GetStore().ListInstalledApps()
+	if err != nil {
+		logger.Error(errors.Wrap(err, "failed to list installed apps"))
+		return false, err
+	}
+
+	// build set of installed app slugs using this license
+	for _, app := range apps {
+		decode := scheme.Codecs.UniversalDeserializer().Decode
+		obj, _, err := decode([]byte(app.License), nil, nil)
+		if err != nil {
+			continue
+		}
+		license := obj.(*kotsv1beta1.License)
+
+		if license.Spec.LicenseID == verifiedLicense.Spec.LicenseID {
+			installedAppSet[app.Slug] = true
+		}
+	}
+
+	// get correct slug for app if it is not installed
+	titleForSlug := strings.Replace(verifiedLicense.Spec.AppSlug, ".", "-", 0)
```

Comment:
why not just get the "not installed yet" apps list using a new function, and then remove the apps that have a license id that match instead of trying to guess the slug here? https://github.com/replicatedhq/kots/pull/2577/files#diff-43a280cac09b9a704ed4f9c88e893f7f46c25ebcb0b26b6b080b40d0599c5efaR48

Reply:
thats a very good suggestion, ill implement that instead, thanks

Reply:
Although I will have to implement a "ListNotInstalledApps" function in a few places to satisfy interface specs